### PR TITLE
Interpolate pointer movements

### LINKED ISSUE #72: Interpolate pointer movements
> This is from the original code; without it, program.wait() waits until the timeout.
> The tests passes (locally) if I program.kill() it instead.

> Ah, I suppose this is because we're moving too far/quickly in one step. I suppose we'll want to interpolate the movements in the `Move Pointer` keyword in smaller increments to avoid this problem.

_Originally posted by @Saviq in https://github.com/MirServer/mir-ci/pull/62#discussion_r1479971101_

### PR DESCRIPTION
**Pull request by hbatagelo**
Fixes #72.

This PR proposes a keyword to incrementally move the pointer from a specified start position to the center of a matching template:

`Walk Pointer From ${point} To Template    ${template}    ${steps}=0    ${step_sleep}=0`

* If `${steps}` <= 0, the number of steps is determined dynamically so that the pointer moves in increments of 16 pixels.
* `${step_sleep}` specifies the time to sleep after each step, with a default of 0.

The drag and drop tests have been updated to use `Walk Pointer` whenever possible, replacing the previous uses of `Displace` followed by `Move Pointer To Absolute`. As the final step of the `Source and Destination Mismatch` test case, the virtual pointer now walks to the "(nothing)" UI label (a label that displays feedback in the match test). This replaces the last `Displace ${center} 20 20` of the test case. The UI label has been enlarged in the client application, and a corresponding template has been created (`drag_and_drop_end.png`). 

To support the `Walk Pointer`, the following auxiliary keywords have been introduced:

* `Distance From ${point_a} To ${point_b}`
* `Lerp Between ${point_a} And ${point_b} With Parameter ${t}`

### CODE REVIEW

**Review by Saviq — COMMENTED** [review]
Hey, this simplifies for me and still works, WDYT?

```diff
diff --git a/mir-ci/mir_ci/test_drag_and_drop.py b/mir-ci/mir_ci/test_drag_and_drop.py
index 139766d..15ea8fd 100644
--- a/mir-ci/mir_ci/test_drag_and_drop.py
+++ b/mir-ci/mir_ci/test_drag_and_drop.py
@@ -89,9 +89,8 @@ class TestDragAndDrop:
             """\
             Source and Destination Match
                 ${pos}=    Move Pointer To Template    ${SRC_TEMPLATE}
-                Sleep    ${A_SHORT_TIME}
                 Press LEFT Button
-                Walk Pointer From ${pos} To Template    ${DST_TEMPLATE}    ${STEPS}    ${A_SHORT_TIME}
+                Walk Pointer From ${pos} To Template    ${DST_TEMPLATE}
                 Release LEFT Button
         """
         )
@@ -124,11 +123,10 @@ class TestDragAndDrop:
             """\
             Source and Destination Mismatch
                 ${pos}=    Move Pointer To Template    ${SRC_TEMPLATE}
-                Sleep    ${A_SHORT_TIME}
                 Press LEFT Button
-                ${pos}=    Walk Pointer From ${pos} To Template    ${DST_TEMPLATE}    ${STEPS}    ${A_SHORT_TIME}
+                ${pos}=    Walk Pointer From ${pos} To Template    ${DST_TEMPLATE}
                 Release LEFT Button
-                Walk Pointer From ${pos} To Template    ${END_TEMPLATE}    ${STEPS}    ${A_SHORT_TIME}
+                Walk Pointer From ${pos} To Template    ${END_TEMPLATE}
         """
         )
 
```

**Comment by hbatagelo** [comment]
It works locally, but does it pass the CI? That would be great (fixes #67)

**Comment by hbatagelo** [comment]
Now that I'm looking at the keywords... notice that some use embedded arguments, such as `Press LEFT Button` instead of `Press Button    LEFT`.
`Walk Pointer` uses a mix of embedded and normal arguments. Maybe we should adopt a single style for consistency?

Embedded arguments improve readability and make the expression more concise. OTOH, optional arguments must always follow the keyword.

With normal arguments it would be something like this:
`Walk Pointer From Position To Template    ${position}    ${template}    ${steps}=0    ${step_sleep}=0`

With embedded arguments (and optional arguments):
`Walk Pointer From ${position} To ${template}    ${steps}=0    ${step_sleep}=0`

**Comment by Saviq** [comment]
> With embedded arguments (and optional arguments): `Walk Pointer From ${position} To ${template} ${steps}=0 ${step_sleep}=0`

I like that better. And yeah, `Press LEFT Button` makes more sense.

Optional arguments are optional, ideally they need not be used (like here - we're down to 4-5 lines).

File: mir-ci/mir_ci/test_drag_and_drop.py [hunk 1]
```diff
@@ -3,25 +3,32 @@
...
+*** Variables ***
+{variables}
+
 *** Test Cases ***
 {test_case}
 """
 ROBOT_SETTINGS = f"Resource   {MIR_CI_PATH}/robot_resources/screencopy.resource"
+ROBOT_VARIABLES = f"""\
+${{A_SHORT_TIME}}    0.3
+${{STEPS}}           3
```
Comment (Saviq) [inline mir-ci/mir_ci/test_drag_and_drop.py:27]: This can go now?
```suggestion
```

File: mir-ci/mir_ci/robot_resources/screencopy.resource [hunk 2]
```diff
@@ -1,23 +1,61 @@
...
 
 Move Pointer To Template
+    [Documentation]    Move the pointer to the center of the first matching image template.
     [Arguments]    ${template}
     ${regions}=    Match    ${template}
     ${center}=    Get Center    ${regions}[0]
     Move Pointer To Absolute    ${center}[x]    ${center}[y]
     RETURN    ${center}
+
+Walk Pointer From ${start_position} To Template
```
Comment (Saviq) [inline mir-ci/mir_ci/robot_resources/screencopy.resource:30]: Should we build required arguments into the keyword then?

```suggestion
Walk Pointer From ${start_position} To Template ${template}
```

File: mir-ci/mir_ci/robot_resources/screencopy.resource [hunk 3]
```diff
@@ -1,23 +1,61 @@
...
+    [Documentation]    Move the pointer in incremental steps from a start position
+    ...    to the center of the first matching image template.
+    ...    If ${steps} <= 0, the pointer is moved in increments of 16 pixels.
+    ...    ${step_sleep} is the time to sleep after each step, in seconds.
+    ...    The return value is the pointer position after the walk.
+    [Arguments]    ${template}    ${steps}=0    ${step_sleep}=0
```
Comment (Saviq) [inline mir-ci/mir_ci/robot_resources/screencopy.resource:36]:
```suggestion
    [Arguments]    ${steps}=0    ${step_sleep}=0
```

**Review by Saviq — APPROVED** [review]
Just one point inline, feel free to merge afterwards.

**Pull request merged**